lburl01/assessment-app
# Feature: Implement UI

#### Background Context
I had from Thursday, March 21 through Monday, March 25th to implement additions on the base assessment app. Most work included making it feasible to update database records via front end forms as well as fetch and display data in the UI to users. 

#### What does this PR do?
- [x] Updates front end dependencies so that Enzyme works with testing and I had access to the relatively new `@untappd/components` `Toggle`
- [x] Adds ability to edit the wins/losses counts for each team within a conference
- [x] Adds ability to view each player within a team 
- [x] Adds ability to edit jersey number and starter values on each player
- [x] Adds migration to add `starter` attr to players
- [x] Adds some model-level validations and relations to ensure no errors trickled through and matched up with the DB schema

#### Screenshots
AFTER: 
<img width="1380" alt="untappd-landing-page" src="https://user-images.githubusercontent.com/19963545/54922383-65d9f580-4ede-11e9-9fd5-cdf9b570c248.png">

![untappd-assessment-gif](https://user-images.githubusercontent.com/19963545/54922391-68d4e600-4ede-11e9-972f-d2be8b1d226e.gif)

#### How should PR reviewers test this?
- Run `yarn setup` to run the migration and install dependencies
- Run `yarn test` to see the new front end tests
- Run `rspec` to see the added backend tests
- Run `yarn start` to fire up the UI; proceed with manual testing

**Ensure you can edit the wins/losses for each team**
- You should see 3 teams under a heading for ACC Conference
- Each team should be in a card element and have a heading with: 
  - Team name (as a link)
  - Mascot
  - Coach's name
  - 2 inputs: 1 for wins, 1 for losses and a submit button
- The wins and losses inputs should be of type number (meaning you can use your keyboard arrow to navigate in steps of `1`)
- Upon submitting a non-whole, non-positive number, you will see a validation error
- Change the value of one of those inputs and submit via the Update button
- The values should persist and you should see a green Toaster success message

**Ensure you can fetch players for each team and then edit player attributes**
- Clicking the team name will trigger a fetch of all players (you can see this in the network tab)
- Each player will be within an individual List Item and display these attributes: 
  - Player name
  - Position
  - Height
  - Weight
  - An input for updating jersey number
  - A toggle for updating whether the player is a starter or not
- Clicking the team name again should hide the player list
- Clicking the team name once more should show players (but not re-fetch the player data)
- Toggling the starter value should persist and show a green Toaster success message
- Change the jersey number to the same number as another player on the same team (input is submitted `onBlur`
  - You should see a red Toaster with a message about uniqueness
  - Now change the jersey to the same as another team's player - the save should succeed and you should see a green Toaster success message

#### Next Steps, Improvements, and Considerations
- I chose not to use Redux for this relatively simple app and was fine passing around state; would this have been part of a larger app with other pages, I would have leveraged Redux more for state management around these components as well has handling interfacing with the backend and extracting the API requests into a separate API file. 
- Typically I add tests for all new code I add; in this case, I focused on one file that I felt was representative of the unit tests I typically write in the front end. I also didn't add any integration/snapshot tests just due to time. 
- There's some funny layout happening with the inline input validation on the jersey; I could either remove the client-side validation and utilize the red Toaster message OR move that input over so that the addition of the error message doesn't adjust the position of the element (I just ran out of time)
